### ISSUE TITLE
requiring sharp breaks ghostscript4js

### ISSUE THREAD

**Issue by Prinzhorn** [lovell/sharp#1244]
This was a wild ride to figure out. We're using ghostscript4js to extract JPEGs from multi-page PDFs. This worked for most files until we had a file where the text of a single PDF would be complete garbage inside the JPEG. For legal reasons I am not able to share the specific PDF, but here's the comparison of PDF vs JPEG

JPEG extracted without sharp required:

![selection_273](https://user-images.githubusercontent.com/679144/40706175-3b94f9fa-63ed-11e8-9ded-c75d6f829d0b.png)

JPEG extract with sharp required (but not used at all):

![selection_274](https://user-images.githubusercontent.com/679144/40706176-3bb62116-63ed-11e8-86cc-5a80ea5ec9df.png)

This is a minimal test-case to reproduce (if you had the PDF...)

```js
//Uncomment this line to break JPEG extraction.
//require('sharp');

const gs = require('ghostscript4js');

async function pdf() {
  try {
    const command =
      '-psconv -q -r300 -sPAPERSIZE=a4 -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -dUseTrimBox -sDEVICE=jpeg -dDownScaleFactor=3 -o out-%d.jpg -dJPEGQ=100 -f in.pdf';
    await gs.execute(command);
    console.log('done');
  } catch (err) {
    console.error(err);
  }
}

pdf();
```

Basically I'm hoping one of the maintainers has an idea as to why simply requiring sharp would break ghostscript? I'm confused af. What exactly does sharp change about the process environment that could cause something like that?

I will spend some time narrowing this down further but I'm hoping one of you knows what's up.

**Comment by Prinzhorn** [comment]
I was able to narrow this down to `0.19.0`. It works with `0.18.4`.

**Comment by lovell** [comment]
Hello, given ghostscript4js depends on globally-installed shared libraries such as libfontconfig you'll need to ensure that a recent-enough version of libvips is also globally-installed on this machine. For sharp v0.20.x you'll need libvips v8.6.1+ - see http://sharp.pixelplumbing.com/en/stable/changelog/

**Comment by Prinzhorn** [comment]
@lovell thank you very much for the timely response. While I try to install the correct version of libvips could you shed some light into why simply requiring sharp would change the process in a way such that a spawned ghostscript process will suddenly use other versions of a shared library?

**Comment by Prinzhorn** [comment]
I believe I understand what is happening and I'd be glad if you could just confirm.

If either of [these two conditions](http://sharp.pixelplumbing.com/en/stable/install/#pre-compiled-libvips-binaries) is met

* If a global installation of libvips that meets the minimum version requirement cannot be found;
* If SHARP_IGNORE_GLOBAL_LIBVIPS environment variable is set.

sharp will not only download a recent version of libvips but also a bunch of other libraries such as libfontconfig and libfreetype (which is sort of documented with "and its dependencies"). I've now installed 8.6.1 manually so that sharp will not download those libs and ghostscript will use the global version of libfontconfig and libfreetype.

Two things are still on my mind:

1. As far as I can tell neither libfontconfig nor libfreetype are a dependency of libvips. Why are they downloaded?
2. I believe the versions that sharp installs are the most recent? If so then sooner or later our PDF will break again if the global version (the apt packages are probably a bit behind) reach that version?

**Comment by lovell** [comment]
> "As far as I can tell neither libfontconfig nor libfreetype are a dependency of libvips. Why are they downloaded?"

libfreetype is a dependency of libfontconfig, which is a dependency of librsvg, which is a dependency of the pre-compiled libvips provided by sharp.

> "I believe the versions that sharp installs are the most recent? If so then sooner or later our PDF will break again if the global version (the apt packages are probably a bit behind) reach that version?"

The pre-compiled binaries of libvips and its dependencies are defined in https://github.com/lovell/sharp-libvips/ and are tagged against the relevant libvips version, e.g. [v8.6.1](https://github.com/lovell/sharp-libvips/blob/v8.6.1/build/lin.sh). When using pre-compiled binaries, the list of versions are also made available at runtime via [sharp.versions](http://sharp.pixelplumbing.com/en/stable/api-constructor/#versions).

It sounds like the underlying problem might be a change in behaviour in libfreetype, either intentional or a regression, between its versions.

**Comment by Prinzhorn** [comment]
Thank you for the explanation
